[lock/unlock] feat: arb L1<->L2 bridge plugin support

makramkd opened the pull request:
This PR implements off-chain bridging support for Arbitrum and for the Mock Bridge being used in the integration tests. In order to properly test this logic in the integration tests a lot of changes unfortunately had to be made and this PR is quite big. This description aims to summarize the changes on each part of the rebalancer codebase so far.

## Bridges
"Bridging support" in the context of the rebalancer plugin involves mainly two things as of time of writing (and may potentially involve more things in the future as we develop this plugin):

1. Determining transfer state: The plugin needs to know the state of each pending transfer that was initiated using a bridge. Certain states allow the plugin to perform certain actions. For example, a pending transfer in the state "finalized" means that the plugin can instruct the Rebalancer contract onchain to inject liquidity into the attached liquidity container.
2. Determining bridge-specific arguments: Each "side" of the bridge involves using different sets of arguments for many bridges, and therefore bridges are directional - e.g L1 -> L2 and L2 -> L1 are different bridge implementations. For example, for the L2 -> L1 direction, the bridge-specific arguments determined for finalizing withdrawals on L1 are fundamentally different than finalizing transfers in the L1 -> L2 direction.

The `Bridge` interface in `core/services/ocr2/plugins/rebalancer/bridge/bridge.go` is the interface that will be implemented for all bridge directions and chains supported. See the interface for the exact methods, which have descriptive comments.

### Arbitrum L1 <-> L2 Bridge Implementations

In this PR, the `Bridge` interface referenced above is implemented for both directions of the Arbitrum One bridge, L1 -> L2 and L2 -> L1. Their contents are mostly based off of the scripts merged in #480, so in essence there is not much new here other than rejigging some things in order to be able to properly implement the logic housed in those scripts.

## Graphs and Discovery

Discovery of the graph of liquidities was tightly coupled to the EVM implementation of the `Rebalancer` interface in the `liquiditymanager` package, and this was causing issues for initializing the bridges, since they also need to know a lot of information about the graph as well, mainly the rebalancer and bridge adapter contract addresses.

As such, this PR factors out the discovery from `liquiditymanager` into a new `discoverer` package. This package provides the `Discoverer` interface which is able to discover rebalancers starting from a master/root/main rebalancer contract and network.

Right now only an EVM implementation is provided, which will be sufficient for our needs, but can be expanded in the future if there are non-EVM rebalancers.

The `liquiditygraph` package was renamed to `graph` since it is now housing much more than just liquidity data:

```golang
type XChainRebalancerData struct {
	RemoteRebalancerAddress   models.Address
	LocalBridgeAdapterAddress models.Address
	RemoteTokenAddress        models.Address
}

type Data struct {
	Liquidity         *big.Int
	TokenAddress      models.Address
	RebalancerAddress models.Address
	XChainRebalancers map[models.NetworkSelector]XChainRebalancerData
}
```

The data housed now closely mirrors what is on-chain, since pretty much all of. this is needed in the off-chain components (e.g bridge implementations).

## Rebalancing Algorithms

The `random` and `dummy` algorithms have been deleted to ease maintenance burdens and only `pingpong` is now available and used within tests.

## Plugin Logic

Plugin logic has been changed in order to account for fetching bridge-specific data. The phases now operate as follows:

1. `Observation`: the plugin syncs the rebalancer graph, by doing a discovery and updating the liquidities on each rebalancer. This behavior was there before and is unchanged. *Additional behavior*: we check the previous outcome and check if there are any _proposed transfers_ from the committed outcome. These outcomes are then _resolved_ by using the bridge implementations. See `resolveProposedTransfers` in `core/services/ocr2/plugins/rebalancer/plugin.go` for more details.
2. `Outcome`: the plugin constructs an outcome by: (a) computing median liquidity per-chain (b) computing the consensus graph of connections between chains (c) consensus on pending transfers. This behavior was there before and is unchanged. *Additional behavior*: we compute a quorum on the _resolved transfers_ that have been sent as part of the observation. This is similar to the pending transfers consensus but involves a little more details. See `computeResolvedTransfersQuorum` in `core/services/ocr2/plugins/rebalancer/plugin.go` for details.
3. `Reports`: the previous plugin logic here was actually buggy. In order to make it clearer what this function is supposed to do the code now determines _incoming_ and _outgoing_ transfers for each network. _Incoming_ transfers are transfers that are meant to generate `Receive` instructions in the report, and _outgoing_ transfers are transfers that are meant to generate `Send` instructions in the report. See `Reports` in `core/services/ocr2/plugins/rebalancer/plugin.go` for more details on this.
4. `ShouldTransmitAcceptedReport`: check the sequence number that has been transmitted onchain and compare it to the provided `seqNr`. If `onchainSeqNr < seqNrArg` then we return `true`, otherwise we return `false`.

makramkd committed:
[lock/unlock] feat: arb L1<->L2 interactions

Add a few functions that will perform Arbitrum briding operations
through the Arbitrum bridge adapters:

* L2 -> L1 withdraw: deploy the arbitrum L2 bridge adapter and perform a
  "withdraw" operation. Tested w/ WETH, however should work for any
  regular ERC-20 supported by the native bridge.
* L1 finalization: deploy the arbitrum L1 bridge adapter and perform a
  "finalize" operation. Tested w/ WETH.

makramkd committed:
update comments, remove unused code

makramkd committed:
Update README.md

makramkd committed:
use gethwrappers instead of json abi's

makramkd committed:
bad push

makramkd committed:
ignore gosec lint

makramkd committed:
feat: fee estimation for arb L1 -> L2

estimate the fees required to reliably transfer funds from L1 to L2 on
arbitrum.

makramkd committed:
lint & test fixes

makramkd committed:
goimports

makramkd committed:
Merge branch 'ccip-develop' into makram/arbl1l2-scripts

makramkd committed:
small fixes

makramkd committed:
simplify l1 -> l2 sending

makramkd committed:
wip

makramkd committed:
more wip

makramkd committed:
more wip, some tests

makramkd committed:
small fixes, more tests

makramkd committed:
wip l1 -> l2 bridge impl

makramkd committed:
impl l1 -> l2 bridge

makramkd committed:
bump snapshot

makramkd committed:
refactor discovery out of the rebalancer type

create a new package solely for discovery to decouple the two

makramkd committed:
init bridge factory in relayer

makramkd committed:
small fix

makramkd committed:
fix tests

makramkd committed:
lint and go generate

makramkd committed:
some renames

makramkd committed:
get pending transfers from the bridges

remove the GetPendingTransfers function from the
liquiditymanager.Rebalancer interface and rely on the bridges to give
the plugin pending transfer states.

makramkd committed:
wip: test bridge impl, incorporating changes from ccip-develop

makramkd committed:
rejig outcome/observations a bit

makramkd committed:
fix tests

switch to our own big type for more predictable JSON encoding

makramkd committed:
make generate

makramkd committed:
expand test to check for sent+finalize combo

makramkd committed:
fixes

makramkd committed:
axe the random rebalancer

use ping pong for integration tests and local don spinups

makramkd committed:
Merge branch 'ccip-develop' into makram/arb-bridge-plugin-support

makramkd committed:
fix NodeInterface compilation

makramkd committed:
revert go.md changes

makramkd commented on contracts/src/v0.8/rebalancer/interfaces/IRebalancer.sol:
```diff
@@ -5,8 +5,14 @@ import {IBridgeAdapter} from "./IBridge.sol";
   struct SendLiquidityParams {
+    /// @notice The amount of tokens to be sent to the remote chain.
     uint256 amount;
+    /// @notice The amount of native that should be sent by the rebalancer in the sendERC20 call.
+    /// @notice This is used to pay for the bridge fees.
+    uint256 nativeBridgeFee;
```
Major change here: the plugin tells the rebalancer how much native to send in the call to the bridge adapter.

makramkd commented on contracts/src/v0.8/rebalancer/bridge-adapters/ArbitrumL2BridgeAdapter.sol:
```diff
@@ -28,6 +28,28 @@ contract ArbitrumL2BridgeAdapter is IBridgeAdapter {
+  );
+  /// @notice event to track the finalization of the L2 to L1 transfer
+  /// @dev no data to emit since there isn't typically a finalization step for L1 to L2 transfers
+  event ArbitrumL1ToL2ERC20Finalized();
```
It would be nice to have a `nonce` here or similar but since this bridge is trustless and can be called by anyone onchain, someone can just call it with a nonce and pretend that a particular tx was finalized, even though it wasn't.

makramkd replied:
Although we could potentially also emit `msg.sender` as well and check that it's the rebalancer contract... @RensR wdyt?

RensR replied:
This is the reason events were in the rebalancer, and not here. If all calls go through there anyway, why add events here? We cannot trust that they are ours

makramkd replied:
Fixed.

makramkd committed:
simplify bridge iface

makramkd commented on core/services/ocr2/plugins/rebalancer/graph/graph.go:
This file is 95% the same as `liquiditygraph/graph.go`, not sure why it's showing up as such a huge diff.

makramkd commented on core/services/ocr2/plugins/rebalancer/graph/graph.go:
```diff
@@ -0,0 +1,311 @@
+	TokenAddress      models.Address
+	RebalancerAddress models.Address
+	XChainRebalancers map[models.NetworkSelector]XChainRebalancerData
+}
```
New additions to the graph data.

makramkd commented on core/services/ocr2/plugins/rebalancer/internal/rebalancer_integration_test.go:
```diff
@@ -61,7 +61,7 @@ var (
 
 func TestRebalancer_Integration(t *testing.T) {
-	newTestUniverse(t, 3)
+	newTestUniverse(t, 2)
```
The tests are easier to debug for now with just 2 chains enabled, we can bump up this number as we squash more bugs.

makramkd commented on core/services/ocr2/plugins/rebalancer/internal/rebalancer_integration_test.go:
```diff
@@ -427,15 +424,23 @@ func waitForTransmissions(
+			sentEvents[sent.Nonce.String()] = struct{}{}
+		case fin := <-erc20FinalizedSink:
+			t.Log("got erc20 finalized event, amount:", fin.Amount, "tx hash:", fin.Raw.TxHash, "nonce:", fin.Nonce)
+			if _, exists := sentEvents[fin.Nonce.String()]; exists {
```
Test logic update here.

makramkd commented on core/services/ocr2/plugins/rebalancer/liquidityrebalancer/rebalancer.go:
```diff
@@ -1,14 +1,14 @@
-	) ([]models.Transfer, error)
+		g graph.Graph,
+		inflightTransfers []models.PendingTransfer,
+	) ([]models.ProposedTransfer, error)
```
Minor interface change here: `Transfer` to `ProposedTransfer` to make things clear these objects cannot be used other than to indicate an intention to transfer.

makramkd commented on core/services/ocr2/plugins/rebalancer/models/models.go:
```diff
@@ -9,7 +9,10 @@ import (
+	"github.com/ethereum/go-ethereum/common/hexutil"
 	chainsel "github.com/smartcontractkit/chain-selectors"
+
+	ubig "github.com/smartcontractkit/chainlink/v2/core/chains/evm/utils/big"
```
Using `ubig` for the more predictable JSON encoding semantics.

makramkd commented on core/services/ocr2/plugins/rebalancer/models/models.go:
```diff
@@ -73,14 +102,34 @@ func NewTransfer(from, to NetworkSelector, amount *big.Int, date time.Time, brid
 type PendingTransfer struct {
 	Transfer
 	Status TransferStatus
+	ID     string
```
Pending transfers can actually have globally unique ID's that will be the same across all oracles - for EVM, they can be (txHash, logIndex), but only after the block that the tx was included in was finalized.

makramkd commented on core/services/ocr2/plugins/rebalancer/models/observation.go:
```diff
@@ -18,20 +20,26 @@ func (n NetworkLiquidity) String() string {
 	LiquidityPerChain []NetworkLiquidity
-	PendingTransfers  []PendingTransfer
-	Edges             []Edge
+	// ResolvedTransfers are the resolved versions of the proposed transfers in the last outcome.
+	ResolvedTransfers []Transfer
```
New field

makramkd commented on core/services/ocr2/plugins/rebalancer/models/observation.go:
```diff
@@ -51,14 +59,31 @@ func DecodeObservation(b []byte) (Observation, error) {
+	// Proposed transfers are only resolved in the Observation stage of OCR3.
+	ProposedTransfers []ProposedTransfer
+
+	// These are transfers that have been proposed by the rebalancing algorithm and have been
+	// resolved in the last observation.
+	// Since these are "send" operations, they are ready to execute onchain.
+	ResolvedTransfers []Transfer
```
New field

makramkd committed:
update tests

makramkd committed:
stub more tests

makramkd committed:
working ShouldTransmitAcceptedReport

makramkd committed:
simplify test bridge logic

makramkd committed:
rm unused func

dimkouv commented on contracts/src/v0.8/rebalancer/Rebalancer.sol:
```diff
@@ -125,22 +127,24 @@ contract Rebalancer is IRebalancer, OCR3Base {
 
   /// @notice Transfers liquidity to another chain.
   /// @dev Called by both the owner and the DON.
   function _rebalanceLiquidity(
```
I think `rebalanceLiquidity` is not a good name for this function, rebalance indicates a transfer in any direction to reach balance.

Maybe something like `moveLiquidity` is better

makramkd replied:
100% agree!

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/arb/contracts.go:
```diff
@@ -0,0 +1,45 @@
+	AllContracts = map[uint64]Contracts{
+		chainsel.ETHEREUM_TESTNET_SEPOLIA_ARBITRUM_1.Selector: {
+			L1: struct {
+				RollupAddress        common.Address
```
should we declare and reuse a struct?

makramkd replied:
Yeah I think that's better, will do that!

makramkd committed:
Merge branch 'ccip-develop' into makram/arb-bridge-plugin-support

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/arb/l2_to_l1.go:
```diff
@@ -0,0 +1,739 @@
+	if len(finalizeInboundTransferMethod.Inputs) != 5 {
+		panic("finalizeInboundTransfer has unexpected number of inputs")
+	}
+	if finalizeInboundTransferMethod.Inputs[0].Type.String() != "address" ||
```
i suggest placing this validations in a function, that way it's also testable

makramkd replied:
Added tests!

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/arb/l2_to_l1.go:
```diff
@@ -0,0 +1,739 @@
+	localSelector       models.NetworkSelector
+	remoteSelector      models.NetworkSelector
+	l1RebalancerAddress common.Address
+	l2BridgeAdapter     *arbitrum_l2_bridge_adapter.ArbitrumL2BridgeAdapter
```
can we make this an interface?

makramkd replied:
Yes, I had the same thought, switching them to interfaces and generating mocks for them 👍

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/arb/l2_to_l1.go:
```diff
@@ -0,0 +1,739 @@
+	lggr                logger.Logger
+	l2Client            client.Client
+	arbSys              *arbsys.ArbSys
+	l2ArbGateway        *l2_arbitrum_gateway.L2ArbitrumGateway
```
same for all this? can we make them interfaces?

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/arb/l2_to_l1.go:
```diff
@@ -0,0 +1,739 @@
+	}
+
+	l1FilterName := fmt.Sprintf("ArbitrumL2ToL1Bridge-L1-%s-%s", remoteChain.Name, localChain.Name)
+	err = l1LogPoller.RegisterFilter(logpoller.Filter{
```
how do we unregister this filters?

-- ignore this if the answer lies below

makramkd replied:
I added a `Close` method. I think it still needs to get called from the plugin

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/arb/l2_to_l1.go:
```diff
@@ -0,0 +1,739 @@
+	// Addresses provided to the below wrappers don't matter,
+	// we're just using them to parse the needed logs easily.
+	l2ArbGateway, err := l2_arbitrum_gateway.NewL2ArbitrumGateway(
+		common.HexToAddress("0x0"),
```
there is a `ZeroAddress()` somewhere i think `utils.ZeroAddress`

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/arb/l2_to_l1.go:
```diff
@@ -0,0 +1,739 @@
+	}, nil
+}
+
+func (l *l2ToL1Bridge) QuorumizedBridgePayload(payloads [][]byte) ([]byte, error) {
```
nit: rename receiver `l` to `b`

makramkd replied:
How come? Struct name starts with an `l`, that's just the rule of thumb I usually follow for member functions and receiver names.

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/arb/l2_to_l1.go:
```diff
@@ -0,0 +1,739 @@
+	// close log poller filters
+	err := l.l2LogPoller.UnregisterFilter(l.l2FilterName)
+	err2 := l.l1LogPoller.UnregisterFilter(l.l1FilterName)
+	return multierr.Combine(err, err2)
```
nit:
```
multierr.Combine(
   unregister(),
   unregister()
)
```

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/arb/l2_to_l1.go:
```diff
@@ -0,0 +1,739 @@
+	l2ToL1Finalizations, err := l.l1LogPoller.LogsCreatedAfter(
+		L2toL1ERC20FinalizedTopic,
+		l.l1BridgeAdapter.Address(),
+		time.Now().Add(-DurationMonth/2),
```
Please add a todo to make this more performant, should be important. The data returned from db will be a lot considering that we get Everything not only our logs

makramkd replied:
Yeah, this looks back for half a month (approximately 15 days) at the moment. The only filter is that the topic is ERC20Finalized. However I think there is room to add a log poller query that allows us to query more than one topic index, I think it'll simplify a lot of the code here that's doing manual filtering.

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/arb/l2_to_l1.go:
```diff
@@ -0,0 +1,739 @@
+		logpoller.Finalized,
+		pg.WithParentCtx(ctx),
+	)
+	// TODO: check if err is sql.ErrNoRows
```
can we check sql.ErrNoRows and remove the todo?

makramkd replied:
Done!

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/arb/l2_to_l1.go:
```diff
@@ -0,0 +1,739 @@
+
+	parsedL2ToL1Finalizations, err := l.parseL2ToL1Finalizations(l2ToL1Finalizations)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse L2 -> L1 finalizations: %w", err)
```
nit: i suggest not using `failed to` in the error msgs `parse l2 -> l1 finalizations` should be enough, it's an error so it already iondicates failure

makramkd replied:
Done.

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/arb/l2_to_l1.go:
```diff
@@ -0,0 +1,739 @@
+	// for the remaining as-of-yet unfinalized transfers, determine if they
+	// are ready to finalize
+	// TODO: edge case: determine finalized but not executed by checking rebalancer's LiquidityTransferred logs
+	ready, readyData, notReady, err := l.partitionReadyTransfers(ctx, unfinalizedTransfers)
```
please add a todo to add debug logs here

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/arb/l2_to_l1.go:
```diff
@@ -0,0 +1,739 @@
+	notReady []*arbitrum_l2_bridge_adapter.ArbitrumL2BridgeAdapterArbitrumL2ToL1ERC20Sent,
+	parsedToLP map[logKey]logpoller.Log,
+) ([]models.PendingTransfer, error) {
+	if len(ready) != len(readyData) {
```
nit: should we create a new data type?
instead of passing both `ready` and `readyData` around
```
type EventWithDataOrWhatever struct {
   event ArbitrumL2BridgeAdapterArbitrumL2ToL1ERC20Sent
   data []byte
}

// and pass one slice
ready []SentFoo
```

makramkd replied:
`ready` and `readyData` are local to this bridge implementation, so it's not passed around further than what you see here (into `toPendingTransfers`). Callers don't see this, only `models.PendingTransfer`s

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/arb/l2_to_l1.go:
```diff
@@ -0,0 +1,739 @@
+			Transfer: models.Transfer{
+				From:   l.localSelector,
+				To:     l.remoteSelector,
+				Amount: ubig.New(transfer.Amount),
```
What's the `ubig` over `big`

makramkd replied:
`ubig` has an explicitly defined `MarshalJSON`. We've been running into a lot of issues with unintuitive JSON marshaling so having a function to refer to in times of need is useful.

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/arb/l2_to_l1.go:
```diff
@@ -0,0 +1,739 @@
+	for _, transfer := range notReady {
+		transfers = append(transfers, models.PendingTransfer{
+			Transfer: models.Transfer{
+				From:   l.localSelector,
```
nit: I suggest initializing using constructors. That way adding a new field to the type raises syntax errors so we don't forget to define the new field

makramkd replied:
I think constructors in Go are not always super useful. In the case of transfer, it has about 10 fields, so the constructor is now a function that takes in 10 arguments. When we add a new field we also need to remember to update the constructor as well (which I actually forgot to do for some of the fields here and Go did not complain because it doesn't really care about constructor functions). Autocomplete is also pretty crappy for functions across all IDE's whereas its much better for struct initializers.

If we want to enforce always passing in all arguments via the compiler itself, we can create the struct using unnamed fields, e.g instead of `models.Transfer{ From: ..., To: ..., ... }` it would be `models.Transfer{fromVal, toVal, ...}` and the compiler will complain if you don't pass in all the struct fields. The research team at chainlink currently uses this paradigm. To make it more readable you can add the field name as a comment right next to the value.

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/arb/l2_to_l1.go:
```diff
@@ -0,0 +1,739 @@
+	var (
+		l2ToL1TxLog, withdrawalInitiatedLog, txToL1Log *gethtypes.Log
+	)
+	for _, lg := range receipt.Logs {
```
should we validate len(lg.Topics) > 0 ?

makramkd replied:
That should never happen but it can't hurt!

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/arb/l2_to_l1.go:
```diff
@@ -0,0 +1,739 @@
+	return params, nil
+}
+
+func unpackUint256(data []byte) (*big.Int, error) {
```
place this in some libs package?

makramkd replied:
I don't think that's needed yet, it's just used by the arbitrum bridge at the moment. Even if it isn't though I would argue for Go's "[A little copying is better than a little dependency.](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=9m28s)" proverb applying here.

dimkouv commented on core/services/ocr2/plugins/rebalancer/bridge/bridge.go:
```diff
@@ -0,0 +1,278 @@
+
+	switch source {
+	case models.NetworkSelector(chainsel.ETHEREUM_MAINNET_ARBITRUM_1.Selector):
+		fallthrough
```
Why? fallthrough will immediately match the next case.
So if source is `chainsel.ETHEREUM_MAINNET_ARBITRUM_1.Selector`
then we execute the code for `models.NetworkSelector(chainsel.ETHEREUM_TESTNET_SEPOLIA_ARBITRUM_1.Selector)`

dimkouv replied:
i see now why but maybe better to do:

```
case A, B:
   do this()
```

instead of

```
case A:
   fallthrough
case B:
   do this()
```

makramkd replied:
Yup this is equivalent to `case A, B` but because the constants are very long I chose to do the `fallthrough` instead, which is easier to read (less screen width used). More akin to classical C switches (which don't have `fallthrough`, they're implicit).

I think what may make this really more readable is if we had this notion of "family" or similar in chain-selectors. E.g all arbitrum mainnets + testnets are the "Arbitrum" family. All ethereum testnets + mainnets are the "Ethereum" family. etc. Then you could have a switch like this:

```golang
switch source.Family() {
case chainsel.Arbitrum:
	if dest.Family() != chainsel.Ethereum {
		return errors.New(...)
	}
case chainsel.Ethereum:
	// ... etc
}
```

makramkd replied:
Oh nevermind actually, I can add a newline after the comma 😄

makramkd committed:
CR review and add more tests

makramkd committed:
some more tests

makramkd committed:
lint

makramkd committed:
more logging, small fixes

makramkd committed:
two important fixes:

* for the arbitrum L2 -> L1 bridge, do not search for finalized
  ERC20Finalized logs but only those that have a single confirmation.
  This is so that we don't mark the same ERC20Sent event as finalized
  more than once, causing on-chain reverts. There might be a better
  solution to this.
* for the arbitrum L1 -> L2 bridge, fix the L2CallValue setting in the
  RetryableData record. This is actually the amount being transferred
  and not a hardcoded "1" value. Updated the scripts to match.

makramkd committed:
add more logging

makramkd committed:
fix tests

makramkd committed:
Merge branch 'ccip-develop' into makram/arb-bridge-plugin-support

RensR commented on contracts/src/v0.8/rebalancer/Rebalancer.sol:
```diff
@@ -34,7 +34,9 @@ contract Rebalancer is IRebalancer, OCR3Base {
     address to,
-    uint256 amount
+    uint256 amount,
+    uint256 bridgeFeePaid,
```
Do we need this in this event?

makramkd replied:
I added this for offchain tracking purposes but maybe this isn't needed for now. I can remove it. It's not checked in tests or in the bridges.

makramkd replied:
Removed.

RensR commented on contracts/src/v0.8/rebalancer/Rebalancer.sol:
```diff
@@ -186,7 +190,12 @@ contract Rebalancer is IRebalancer, OCR3Base {
     // finalize the withdrawal through the bridge adapter
     // TODO: handle edge case where withdrawal is done already
-    // TODO: remoteRebalancer.localBridge.finalizeWithdrawERC20(amount, bridgeData);
+    // maybe just try/catch is enough?
```
Try catch?

makramkd replied:
Gonna leave this here for now, there might be pros and cons to reverting eagerly in the finalize vs. soft reverting with the try/catch. Can revisit after this PR.

makramkd replied:
Thinking on this some more, I think try/catch is sufficient.

makramkd replied:
Thinking on this some more, I think try/catch is sufficient.

RensR commented on contracts/src/v0.8/rebalancer/bridge-adapters/ArbitrumL1BridgeAdapter.sol:
```diff
@@ -97,20 +125,25 @@ contract ArbitrumL1BridgeAdapter is IBridgeAdapter {
-        abi.encode(params.maxSubmissionCost, bytes(""))
-      );
+    // outboundTransferCustomRefund will return the abi encoded inbox sequence number
+    // which is 256 bits, so we can cap the return data to 256 bits, e.g using https://github.com/nomad-xyz/ExcessivelySafeCall
```
I think we're good on trusting the bridge to not return bomb us

makramkd replied:
Sounds good, will remove these comments 🙏

makramkd replied:
Removed.

RensR commented on contracts/src/v0.8/rebalancer/bridge-adapters/ArbitrumL1BridgeAdapter.sol:
```diff
@@ -53,6 +53,34 @@ contract ArbitrumL1BridgeAdapter is IBridgeAdapter {
+  /// @dev contracts. And since the official bridge contracts probably have lots of
+  /// @dev events and logs, we can use this event to track the L2 to L1 transfers
+  /// @dev with less load on the rebalancer oracles.
+  event ArbitrumL2ToL1ERC20Finalized(
```
Not sure I see a lot of value in events here over the permissioned rebalancer

makramkd replied:
The main reason this is here is that in theory someone could finalize the transfer through the B.A before the rebalancer finalizes it.

Right now we don't fully protect against this "attack" because someone could technically finalize without the bridge adapter and call the outbox's `executeTransaction`. To protect fully we'd have to watch for `OutBoxTransactionExecuted` on the outbox itself but that is probably a high volume log (though we could have low retention on this log in the log poller).

This is a good thing to address in a follow up PR because the bridge logic that determines a pending transfer's state should be bulletproof. But if we want to address this in this PR can also do that, just will push out merging it for a bit while we implement and test the logic.

makramkd replied:
After thinking about it some more I'll cut over to the LiquidityTransferred events exclusively and use OutBoxTransactionExecuted to handle this edge case in a future PR. I think it's cleaner to handle the edge case once and for all rather than partially.

makramkd replied:
Removed these events.

RensR commented on contracts/src/v0.8/rebalancer/bridge-adapters/ArbitrumL2BridgeAdapter.sol:
```diff
@@ -49,17 +71,27 @@ contract ArbitrumL2BridgeAdapter is IBridgeAdapter {
+
+    emit ArbitrumL2ToL1ERC20Sent(localToken, remoteToken, recipient, amount, res);
+
+    return res;
```
Can we use full words?

makramkd replied:
Made variable names more clear.

RensR commented on contracts/src/v0.8/rebalancer/test/Rebalancer.t.sol:
```diff
@@ -111,19 +115,19 @@ contract Rebalancer_rebalanceLiquidity is RebalancerSetup {
 
     deal(address(s_l1Token), address(s_bridgeAdapter), amount);
 
-    s_rebalancer.rebalanceLiquidity(i_remoteChainSelector, amount, bytes(""));
```
Feels like the test case there it isn't zero is missing here

makramkd replied:
Good note, will add.

makramkd replied:
Added a test case below!

Commit:
rebalancer - config digests consensus (#531)

Come in consensus on config digests.

makramkd committed:
address CR comments

makramkd committed:
fix snapshot

makramkd committed:
use only LiquidityTransferred events

makramkd committed:
fix tests

makramkd committed:
revert logging changes

makramkd committed:
revert logging changes

makramkd committed:
revert logging changes

makramkd committed:
address more PR comments

* add try/catch in Rebalancer's finalize call
* simulate an "already finalized" situation in the mock bridge adapter
  by reverting on an already used nonce.

makramkd committed:
add error event in failed finalization case

makramkd committed:
Merge branch 'ccip-develop' into makram/arb-bridge-plugin-support

makramkd committed:
add coverage for finalization failed case

makramkd committed:
prettier

RensR commented on contracts/src/v0.8/rebalancer/test/Rebalancer.t.sol:
```diff
@@ -135,6 +143,109 @@ contract Rebalancer_rebalanceLiquidity is RebalancerSetup {
+    // deal some L2 tokens to the remote token pool so that we can withdraw it when we rebalance.
+    deal(address(s_l2Token), address(remotePool), amount);
+
+    vm.expectEmit();
```
nit we call expectEmit only immediately before the emit

makramkd replied:
Nice catch, done.

RensR commented on contracts/src/v0.8/rebalancer/test/Rebalancer.t.sol:
```diff
@@ -135,6 +143,109 @@ contract Rebalancer_rebalanceLiquidity is RebalancerSetup {
+    assertEq(s_l1Token.balanceOf(address(s_bridgeAdapter)), 0, "bridgeAdapter balance");
+
+    // try to finalize on L1 again
+    vm.expectEmit();
```
This expect should be before the emit, and you need to expect 2 more times to assert on the other 2 below

makramkd replied:
Good bug catch, fixed.

makramkd committed:
fix test assertions

makramkd committed:
make receiveLiquidity onlyOwner
